### ISSUE TITLE
Protect against DNS hijacking by node

### ISSUE THREAD

**Issue by cvl** [mysteriumnetwork/node#3227]
Related to question: https://github.com/mysteriumnetwork/node/issues/3223

Apparently node can hijack DNS and send the wrong (e.g. phishing) IP address to consumer (similar to https://mobile.twitter.com/CreamdotFinance/status/1371448627663491088).

We need to find a way to make sure this is not possible or nearly impossible, or either a) revert back to centralized DNS resolution solution, b) have certain amount of trusted & carefully vetted + KYC'ed nodes which can be used for DNS resolution.

This one is critical as at least one outbreak can lead to collapse of whole project.

CC: @chompomonim

**Comment by cvl** [comment]
Yes

**Comment by Snawoot** [comment]
Hello!

> Apparently node can hijack DNS and send the wrong (e.g. phishing) IP address to consumer (similar to https://mobile.twitter.com/CreamdotFinance/status/1371448627663491088).

In their case they have lost control over DNS servers, had them compromised. Malicious actors seized same capabilities as domain legitimate owners and were able to do everything with domain, including issuance of TLS (SSL) certificates for domain because they could prove domain ownership to certificate authorities, issuing certificates automatically. No amount of client side security will fix this because past this point attackers define authoritative DNS server responses. This case is rather unrelated to traffic spoof in the middle of some consumer connection: key difference is de-facto domain ownership and ability to issue certificates.

Speaking of consumer traffic hijacking, exit node owner in Mysterium has exactly same capabilities as any ISP or Tor node operator or any VPN service: they are able to modify and divert users' traffic in arbitrary destination and see it in the same way as it leaves your machine before entering encrypted connection. Tunnel encryption here adds security only against other observers between consumer and provider. End-to-end confidentiality and authenticity still has to be ensured by, for example, TLS connection between user and accessed Internet resource.

> We need to find a way to make sure this is not possible or nearly impossible, or either a) revert back to centralized DNS resolution solution, b) have certain amount of trusted & carefully vetted + KYC'ed nodes which can be used for DNS resolution.

Most important part, DNS security here is irrelevant. It is possible (and even more convenient option) to hijack traffic using off-the-shelf transparent proxies, without even changing IP addresses for domain, perceived by consumer. For example, if HTTP connection is not protected by TLS (HTTPS), ISP/Tor node operator/VPN service/Mysterium node owner needs only one packet filter rule to redirect it to another destination, rewriting IP addresses back and forth, and keeping client convinced they're talking to destination they requested.

Regarding option A you've presented, even plain requests DNS can be subject of redirection, so it doesn't changes anything even for DNS itself.

In fact, ISPs and VPN providers are using such techniques at scale. ISPs use it to enforce restrictions and VPN providers use it to snoop into SNI, determine destination domain and divert users traffic to residential exit nodes in order to bypass restrictions, while rest of traffic originates from cheap datacenter channels.

Simply put, any VPN/Tor/Mysterium/whatever is not a replacement for other end-to-end security. It just shifts your exit point to another destination and makes sure that shift was done securely over all middleboxes on the way. Same rules apply as usual past tunnel exit, and node operator is yet another middlebox on the way from exit node to actual destination.

Answer provided by @soffokl in [your previous issue](https://github.com/mysteriumnetwork/node/issues/3223) was terse but explains situation in a nutshell. VPN/Tor/Mysterium/any other tunnel doesn't improve any security beyond extent where it operates.

**Comment by cvl** [comment]
Hi @Snawoot 

Thank you for the sophisticated answer.

Just to be clear, did I understand correctly, at this moment, provider can execute the following scenario:
- consumer enters https://www.mybank.com in his browser
- provider redirects this and following requests to his servers (where UI looks-a-like to original mybank.com)
- consumer enters all the passwords etc to this fake https://www.mybank.com

Is this correct or I'm missing something?

**Comment by Snawoot** [comment]
@cvl

You're missing that provider won't be able to have correct HTTPS certificate for www.mybank.com. If he'll redirect connections to own server with inappropriate cert, page in browser will look like this:

https://untrusted-root.badssl.com/

or like this:

https://self-signed.badssl.com/

**Comment by cvl** [comment]
I can see now that all browsers have a pre-installed list of all major CAs public keys.

Thanks!
